compiler/x/codegen/BinaryEvaluator.cpp

```diff
@@ -1933,28 +1933,90 @@ TR::Register *OMR::X86::TreeEvaluator::integerDualMulEvaluator(TR::Node *node, T
 
 TR::Register *OMR::X86::TreeEvaluator::integerMulEvaluator(TR::Node *node, TR::CodeGenerator *cg)
    {
-   TR::Register *targetRegister = 0;
-   TR::Node     *firstChild     = node->getFirstChild();
-   TR::Node     *secondChild    = node->getSecondChild();
-   bool          nodeIs64Bit    = TR::TreeEvaluator::getNodeIs64Bit(node, cg);
-   intptrj_t    value;
+   TR::Register *targetRegister  = 0;
+   TR::Node     *firstChild      = node->getFirstChild();
+   TR::Node     *secondChild     = node->getSecondChild();
+   bool          nodeIs64Bit     = TR::TreeEvaluator::getNodeIs64Bit(node, cg);
+   bool          isLoadConst     = secondChild->getOpCode().isLoadConst();
+   bool          isByteOperation = false;
+   bool          isMemRef        = false;
+   TR_X86OpCodes opCode          = BADIA32Op; 
+   TR_X86OpCodes regOpCode       = BADIA32Op; 
+   TR_X86OpCodes memOpCode       = BADIA32Op; 
+   TR_X86OpCodes movOpCode       = BADIA32Op;
+   int64_t       value;
 
    if (node->isDualCyclic())
       {
       return TR::TreeEvaluator::integerDualMulEvaluator(node, cg);
       }
 
-   if (secondChild->getOpCode().isLoadConst() &&
-      (value = TR::TreeEvaluator::integerConstNodeValue(secondChild, cg)))
+   // Opcode evaluation for bmul, smul imul and lmul
+   switch (secondChild->getDataType())
+      {
+      case TR::Int8:
+         isMemRef = secondChild->getOpCode().isMemoryReference(); 
+         opCode = isMemRef ? IMUL1AccMem : IMUL1AccReg;
+     
+         regOpCode = IMUL1AccReg;
+         memOpCode = IMUL1AccMem;
+         movOpCode = MOV1RegReg;
+
+         isByteOperation = true;
+         break;
+
+      case TR::Int16:
+         if(isLoadConst)
+            {
+            value = secondChild->getShortInt();
+            isMemRef = firstChild->getOpCode().isMemoryReference(); 
+            if (isMemRef)
+               opCode = (value >= -128 && value <= 127) ? IMUL2RegMemImms : IMUL2RegMemImm2;
+            else
+               opCode = (value >= -128 && value <= 127) ? IMUL2RegRegImms : IMUL2RegRegImm2;
+            }
+
+         regOpCode = IMUL2RegReg;
+         memOpCode = IMUL2RegMem;
+         movOpCode = MOV2RegReg;
+         break;
+
+      case TR::Int32:
+      case TR::Int64:
+         if(isLoadConst)
+            {
+            value = TR::TreeEvaluator::integerConstNodeValue(secondChild, cg);
+            // Assign opcodes with Immidiates only if value can be assigned in 4 bytes
+            if(value <= TR::getMaxSigned<TR::Int32>())  
```

Comment:
you need to check value >= MinSigned and value <= MaxSigned - this check is not sufficient.

Reply:
@andrewcraik I have checked in this additional check.